### PR TITLE
ISSUES-995 support relative path with domain

### PR DESCRIPTION
**Pull request by zhang2014**
#995
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/tests/queries/0_stateless/00398_url_functions.sql [hunk 1]
```diff
@@ -1,21 +1,107 @@
...
+SELECT decodeURLComponent(queryStringAndFragment('http://127.0.0.1/'));
+SELECT decodeURLComponent(queryStringAndFragment('http://127.0.0.1/?'));
+SELECT decodeURLComponent(queryStringAndFragment('http://127.0.0.1/?query=hello%20world+foo%2Bbar'));
+SELECT decodeURLComponent(queryStringAndFragment('http://127.0.0.1/?query=hello%20world+foo%2Bbar#a=b'));
+SELECT decodeURLComponent(queryStringAndFragment('http://paul@127.0.0.1/?query=hello%20world+foo%2Bbar#a=b'));
+SELECT decodeURLComponent(queryStringAndFragment('//paul@127.0.0.1/?query=hello%20world+foo%2Bbar#a=b'));
+
+SELECT '====CUT TO FIRST SIGNIFICANT SUBDOMAIN====';
+SELECT cutToFirstSignificantSubdomain('http://www.example.com');
+SELECT cutToFirstSignificantSubdomain('http://www.example.com:1234');
```
Comment (zhang2014) [inline dbms/tests/queries/0_stateless/00398_url_functions.sql]: The result is `example.com`, I'm not sure it's correct.

Comment (zhang2014) [reply dbms/tests/queries/0_stateless/00398_url_functions.sql]: `cutToFirstSignificantSubdomain('http://www.example.com:1234') = 'example.com:1234';` - It's right?

File: dbms/src/Functions/FunctionsURL.h [hunk 2]
```diff
@@ -88,34 +88,38 @@ inline StringView getURLScheme(const StringView & url)
...
-        p += 3;
 
-    const char * st = p;
+    if (pos != url.data())
+    {
+        StringView scheme = getURLScheme(url);
+        Pos scheme_end = url.data() + scheme.size();
+
+        // Colon must follows after scheme.
+        if (*(scheme_end++) != ':' || scheme_end != pos)
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply dbms/src/Functions/FunctionsURL.h]: This also assumes zero terminated string.
(The assumption is actually correct in this case, but it's bad practice.)

File: dbms/src/Functions/FunctionsURL.h [hunk 3]
```diff
@@ -88,34 +88,38 @@ inline StringView getURLScheme(const StringView & url)
...
+            return StringView();
+    }
+
+    if (end - pos < 2 || *(pos++) != '/' || *(pos++) != '/')
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsURL.h]: Unclear.

File: dbms/src/Functions/FunctionsURL.h [hunk 4]
```diff
@@ -392,18 +396,39 @@ struct ExtractWWW
...
+            {
+                if (*pos == '@')
+                {
+                    st = pos + 1;
+                } else if (*pos == ':' || *pos == '/' || *pos == '?' || *pos == '#')
+                {
+                    break;
+                }
+            }
+
```
[the remark this answers is not in the document]
Comment (zhang2014) [reply dbms/src/Functions/FunctionsURL.h]: 😄I've changed it, but I can't fold this comment.

**Comment by zhang2014** [comment]
Maybe we should continue : )

**Comment by alexey-milovidov** [comment]
Yes, this is very useful.
The only missing part is performance tests.

File: dbms/tests/queries/0_stateless/00398_url_functions.sql (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/tests/queries/0_stateless/00398_url_functions.sql]: This is correct.

Domain doesn't contain port number.
The behaviour of `cutToFirstSignificantSubdomain` is consistent to `domain` and `domainWithoutWWW` in this sense.

**Pull request merged**